src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,836 @@
+                            RenewedLease = true;
+                            DrainedLease = true;
+                        }
+                        
+                        try
+                        {
+                            await this.partitionTable.UpdateEntityAsync(partition, etag, (TableUpdateMode)1);
+                            this.LogHelper(partition, false, false, RenewedLease, DrainedLease, ReleasedLease, "");
+                            
+                        }
+                        // Catch exception of failed table update due to ETag expired. 
+                        // Re-read the table immediately to get the latest ETag.
+                        catch (RequestFailedException ex) when (ex.Status == 412)
+                        {
+                            this.settings.Logger.PartitionManagerInfo(
+                                this.storageAccountName,
+                                this.settings.TaskHubName,
+                                this.settings.WorkerId,
+                                partition.RowKey,
+                                $"Failed to update {partition.RowKey} due to concurrency. Re-read the table to get the latest Etag.");
+                            throw;
+                        }
+                        // Eat any other exceptions.
+                        catch (Exception exception)
+                        {
+                            this.settings.Logger.PartitionManagerError(
+                                this.storageAccountName,
+                                this.settings.TaskHubName,
+                                this.settings.WorkerId,
+                                partition.RowKey,
+                                $"Unexpected error occurred in shutdown when updating lease {partition.RowKey} : {exception}"
+                                );
+                        }
+                    }
+
+                }
+                bool releasedAllLeases = (ownershipLeaseCount == 0);
+                return releasedAllLeases;
+            }
+
+           
+        }
+
+        /// <summary>
+        ///The Response class describes the behavior of the ReadandWrite method in the PartitionManager worker class. 
+        ///If the virtual machine is about to drain and release a lease, the method sets the WorkonDrain flag to true. 
+        ///If the VM is going to acquire another lease from other VM, it sets the waitforPartition flag to true. 
+        ///When either of these flags is true, the sleep time of the VM changes to 1 second.
+        /// </summary>
+        class ReadTableReponse
+        {
+            //If set to true, it indicates that the VM is working on release lease. 
+            public bool WorkOnDrain { get; set; } = false;
+            
+            //If set to true, it indicates that the VM is waiting for a lease to be released.
+            public bool WaitForPartition { get; set; } = false;
```

Comment:
Here's a better name suggestion to make the code more understandable:
```suggestion
            public bool IsWaitingToStealPartition { get; set; } = false;
```
Another option might be `IsWaitingForPartitionRelease` but I think I like my first suggestion a little better.

Reply:
Thanks! Here we want the parameter indicatingthe worker waits for the partition which it stole to be released. The partition is already stolen, we don't need to wait to steal, but wait it to be released so to claim it. So, I think `isWaitingForPartitionReleases` is more accurate here?

Reply:
In that context, I'd agree that `isWaitingForPartitionReleases` is a better name :)

Reply:
Oh, sorry I misunderstood it. I will change it to `IsDrainingPartition`. Sorry for the confusion and please ignore my reply above. :)